sydent/http/servlets/__init__.py

```diff
@@ -163,45 +163,27 @@ def inner(self, request: Request, *args, **kwargs) -> bytes:
 
 
 def asyncjsonwrap(f):
-    def reqDone(resp: Dict[str, Any], request: Request) -> None:
-        """
-        Converts the given response content into JSON and encodes it to bytes, then
-        writes it as the response to the given request with the right headers.
-
-        :param resp: The response content to convert to JSON and encode.
-        :param request: The request to respond to.
-        """
-        request.setHeader("Content-Type", "application/json")
-        request.write(dict_to_json_bytes(resp))
-        request.finish()
-
-    def reqErr(failure: Failure, request: Request) -> None:
-        """
-        Logs the given failure. If the failure is a MatrixRestError, writes a response
-        using the info it contains, otherwise responds with 500 Internal Server Error.
-
-        :param failure: The failure to process.
-        :param request: The request to respond to.
-        """
+    async def render(f, servlet, request, **kwargs):
         request.setHeader("Content-Type", "application/json")
-        if failure.check(MatrixRestError) is not None:
-            request.setResponseCode(failure.value.httpStatus)
+        try:
+            result = await f(servlet, request, **kwargs)
+            request.write(dict_to_json_bytes(result))
+            request.finish()
+        except MatrixRestError as e:
+            request.setResponseCode(e.value.httpStatus)
             request.write(
-                dict_to_json_bytes(
-                    {"errcode": failure.value.errcode, "error": failure.value.error}
-                )
-            )
-        else:
-            logger.error(
-                "Request processing failed: %r, %s", failure, failure.getTraceback()
+                dict_to_json_bytes({"errcode": e.value.errcode, "error": e.value.error})
             )
+        except Exception:
+            f = failure.Failure()
+            logger.error("Request processing failed: %r, %s", f, f.getTraceback())
             request.setResponseCode(500)
             request.write(
                 dict_to_json_bytes(
                     {"errcode": "M_UNKNOWN", "error": "Internal Server Error"}
                 )
             )
-        request.finish()
+            request.finish()
```

Comment:
We need to call `request.finish()` no matter what, so we should dedent this (and remove it from above).